### Ejemplo de ajuste de curvas

Para determinar la precisión de un instrumento EDM, se calibró en una línea base de calibración gubernamental. Las distancias medidas corregidas por la condición meteorológica y reducidas de marca a marca y las distancias publicadas correspondientes proporcionadas para la línea base se dan en la Tabla 6.23.

#### Tabla 6.23 Distancias publicadas y medidas de marca a marca para un instrumento EDM.

| Línea | Distancia publicada de marca a marca (m) | Distancia medida de marca a marca (m) | Desviación estándar (m) |
|-------|-----------------------------------------|--------------------------------------|-------------------------|
| 1_2   | 70.0002                                 | 70.0020                              | 0.00101                 |
| 1_3   | 219.9998                                | 219.9982                             | 0.00103                 |
| 1_4   | 670.0132                                | 670.0139                             | 0.00121                 |
| 1_5   | 970.0067                                | 970.0070                             | 0.00140                 |
| 1_6   | 1469.9896                               | 1469.9906                            | 0.00179                 |
| 2_3   | 149.9996                                | 149.9985                             | 0.00102                 |
| 2_4   | 600.0130                                | 600.0118                             | 0.00110                 |
| 2_5   | 900.0066                                | 900.0036                             | 0.00112                 |
| 2_6   | 1399.9894                               | 1399.9921                            | 0.00173                 |
| 3_4   | 450.0134                                | 450.0116                             | 0.00111                 |
| 3_5   | 750.007                                 | 750.0046                             | 0.00126                 |
| 3_6   | 1249.9898                               | 1249.9868                            | 0.00161                 |
| 4_5   | 299.9936                                | 299.9925                             | 0.00105                 |
| 4_6   | 799.9765                                | 799.9759                             | 0.00129                 |
| 5_6   | 499.9829                                | 499.9839                             | 0.00113                 |

#### Tabla 6.23. Una línea de regresión lineal en forma de $ y = a + bx $ se ajustará al conjunto de datos con $ y $ representando las mediciones de distancia corregidas, $ x $ el vector de distancias publicadas (consideradas como valores constantes), y $ a $ y $ b $ los parámetros desconocidos a determinar. Dadas las desviaciones estándar de las mediciones en la última columna de la Tabla 6.23, determina los parámetros ajustados de mínimos cuadrados $ a $ y $ b $, y muestra la gráfica de los conjuntos de datos que muestra la línea de regresión ajustada. (Asume que los valores aproximados de los parámetros son $ a^0 = 0, b^0 = 1 $).

**Solución:**

Formular las ecuaciones del modelo paramétrico ($ \ell = f(x) $) (ignorando los sombreros) de la siguiente manera:

$$
y_1 = a + b(70.0002)
$$
$$
y_2 = a + b(219.9998)
$$
$$
y_3 = a + b(670.0132)
$$
$$
y_4 = a + b(970.0067)
$$
$$
y_5 = a + b(1469.9896)
$$
$$
y_6 = a + b(149.9996)
$$
$$
y_7 = a + b(600.013)
$$
$$
y_8 = a + b(900.0066)
$$
$$
y_9 = a + b(1399.9894)
$$
$$
y_{10} = a + b(450.0134)
$$
$$
y_{11} = a + b(750.007)
$$
$$
y_{12} = a + b(1249.9898)
$$
$$
y_{13} = a + b(299.9936)
$$
$$
y_{14} = a + b(799.9765)
$$
$$
y_{15} = a + b(499.9829)
$$

La primera matriz de diseño ($ A $) y el vector de cierre ($ w $) utilizando los valores aproximados de los parámetros ($ a^0 = 0, b^0 = 1 $):

$$
A = 
\begin{bmatrix}
1 & 70.0002 \\
1 & 219.9998 \\
1 & 670.0132 \\
1 & 970.0067 \\
1 & 1469.9896 \\
1 & 149.9996 \\
1 & 600.0130 \\
1 & 900.0066 \\
1 & 1399.9894 \\
1 & 450.0134 \\
1 & 750.0070 \\
1 & 1249.9898 \\
1 & 299.9936 \\
1 & 799.9765 \\
1 & 499.9829
\end{bmatrix}
$$

$$
w = f(x^0) - \ell = 
\begin{bmatrix}
-0.0018 \\
0.0016 \\
-0.0007 \\
-0.0003 \\
-0.0010 \\
0.0011 \\
0.0012 \\
0.0030 \\
-0.0027 \\
0.0018 \\
0.0024 \\
0.0030 \\
0.0011 \\
0.0006 \\
-0.0010
\end{bmatrix}
$$

### Matriz de pesos $(P)$ de las mediciones como las inversas de las varianzas de las mediciones:
$$ 
P = \text{diag}(10.0, 10.0, 6.944, 5.102, 3.086, 10.0, 6.944, 5.102, 3.460, 8.264, 5.917, 3.906, 8.264, 5.917, 8.264) \times 1E+05
$$

### La matriz de los coeficientes de las ecuaciones normales $(N)$ y el vector de correcciones $(\delta)$ a los parámetros aproximados $(x^0)$:
$$ 
N = A^T P A = 
\begin{bmatrix}
1.0117356E+7 & 5.65265332E+9 \\
5.65265332E+9 & 4.6462173E+12
\end{bmatrix} 
$$
$$ 
u = A^T P w = 
\begin{bmatrix}
5.89914607E+3 \\
5.37565018E+6
\end{bmatrix} 
$$
$$ 
\delta = -N^{-1} u = 
\begin{bmatrix}
-5.531E-3 \\
-5.359E-8
\end{bmatrix} 
$$

### Parámetros ajustados:
$$ 
\hat{x} = 
\begin{bmatrix}
a \\
b
\end{bmatrix} 
$$
$$ 
\hat{x} = x^0 + \delta = 
\begin{bmatrix}
0.0 \\
1.0
\end{bmatrix} 
+ 
\begin{bmatrix}
-5.531E-4 \\
-5.359E-8
\end{bmatrix} 
\rightarrow 
\hat{x} = 
\begin{bmatrix}
-0.55 \ \text{mm} \\
0.9999999464
\end{bmatrix} 
$$

### Vector residual $(v)$: $v = A\delta + w$
$$ 
v = 
\begin{bmatrix}
-2.4 \\
1.0 \\
-1.3 \\
-0.9 \\
-1.6 \\
0.5 \\
0.6 \\
2.4 \\
-3.3 \\
1.2 \\
1.8 \\
2.4 \\
0.5 \\
0.0 \\
-1.6
\end{bmatrix} \times E-3 
$$

### Factor de varianza a posteriori de peso unitario:
$$ 
\sigma_0^2 = \frac{v^T P v}{df} = \frac{24.01632}{13} = 1.8474
$$

### Matriz de covarianza de los parámetros:
$$ 
C_{\hat{x}} = \sigma_0^2 N^{-1} = 
\begin{bmatrix}
5.7014E-7 & -6.94E-10 \\
-6.94E-10 & 1.24E-12
\end{bmatrix} 
$$

In [1]:
import sympy as sp

# Definir las variables
a, b = sp.symbols('a b')

# Formar las ecuaciones paramétricas
y = [
    a + 70.0002 * b,
    a + 219.9998 * b,
    a + 670.0132 * b,
    a + 970.0067 * b,
    a + 1469.9896 * b,
    a + 149.9996 * b,
    a + 600.0130 * b,
    a + 900.0066 * b,
    a + 1399.9894 * b,
    a + 450.0134 * b,
    a + 750.0070 * b,
    a + 1249.9898 * b,
    a + 299.9936 * b,
    a + 799.9765 * b,
    a + 499.9829 * b
]

# Evaluar la Jacobiana con respecto a a y b
J = sp.Matrix(y).jacobian([a, b])
A = J

# Datos proporcionados
L = sp.Matrix([
    70.0020, 219.9982, 670.0139, 970.0070, 1469.9906,
    149.9985, 600.0118, 900.0036, 1399.9921, 450.0116,
    750.0046, 1249.9868, 299.9925, 799.9759, 499.9839
])

# Valores aproximados
a0, b0 = 0, 1

# Calcular L0
L0 = sp.Matrix(y).subs({a: a0, b: b0})

# Vector de cierre w
w = L0 - L

# Formar la matriz de pesos P
P = sp.diag(*[1/std**2 for std in [
    0.00101, 0.00103, 0.00121, 0.00140, 0.00179,
    0.00102, 0.00110, 0.00112, 0.00173, 0.00111,
    0.00126, 0.00161, 0.00105, 0.00129, 0.00113
]])

# Calcular N y u
N = A.T * P * A
u = A.T * P * w

# Solución para δ
d = -N.inv() * u

# Parámetros ajustados
a_adj = a0 + d[0]
b_adj = b0 + d[1]

# Residuales y varianza a posteriori
v = A * d + w
S02 = (v.T * P * v)[0] / (len(L) - 2)

# Matriz cofactor Qx y matriz de covarianza Cx
Qx = N.inv()
Cx = S02 * Qx

# Mostrar resultados
A, w, N, u, d, a_adj.evalf(), b_adj.evalf(), S02, Qx, Cx


(Matrix([
 [1,   70.0002],
 [1,  219.9998],
 [1,  670.0132],
 [1,  970.0067],
 [1, 1469.9896],
 [1,  149.9996],
 [1,   600.013],
 [1,  900.0066],
 [1, 1399.9894],
 [1,  450.0134],
 [1,   750.007],
 [1, 1249.9898],
 [1,  299.9936],
 [1,  799.9765],
 [1,  499.9829]]),
 Matrix([
 [  -0.0017999999999887],
 [  0.00159999999999627],
 [-0.000700000000051659],
 [-0.000299999999924694],
 [-0.000999999999976353],
 [  0.00109999999997967],
 [  0.00120000000003984],
 [  0.00300000000004275],
 [ -0.00270000000000437],
 [  0.00180000000000291],
 [  0.00239999999996598],
 [  0.00300000000015643],
 [  0.00110000000000809],
 [ 0.000599999999963075],
 [  -0.0010000000000332]]),
 Matrix([
 [10465535.8385753, 5976884878.38873],
 [5976884878.38873, 4914710401773.06]]),
 Matrix([
 [7043.95750838673],
 [ 4357900.1594525]]),
 Matrix([
 [-0.000545594742432017],
 [  -2.2319589640245e-7]]),
 -0.000545594742432017,
 0.999999776804104,
 1.94281099824169,
 Matrix([
 [  3.12801289649711e-7, -3.80404366730005e-10],
 

Aquí tienes los resultados de los cálculos en Python utilizando `sympy` para el ajuste de mínimos cuadrados:

### Matriz de diseño $ A $
$$
A = 
\begin{bmatrix}
1 & 70.0002 \\
1 & 219.9998 \\
1 & 670.0132 \\
1 & 970.0067 \\
1 & 1469.9896 \\
1 & 149.9996 \\
1 & 600.013 \\
1 & 900.0066 \\
1 & 1399.9894 \\
1 & 450.0134 \\
1 & 750.007 \\
1 & 1249.9898 \\
1 & 299.9936 \\
1 & 799.9765 \\
1 & 499.9829
\end{bmatrix}
$$

### Vector de cierre $ w $
$$
w = 
\begin{bmatrix}
-0.0018 \\
0.0016 \\
-0.0007 \\
-0.0003 \\
-0.0010 \\
0.0011 \\
0.0012 \\
0.0030 \\
-0.0027 \\
0.0018 \\
0.0024 \\
0.0030 \\
0.0011 \\
0.0006 \\
-0.0010
\end{bmatrix}
$$

### Matriz de los coeficientes de las ecuaciones normales $ N $
$$
N = 
\begin{bmatrix}
1.0465536 \times 10^7 & 5.976884878 \times 10^9 \\
5.976884878 \times 10^9 & 4.914710401773 \times 10^{12}
\end{bmatrix} 
$$

### Vector $ u $
$$
u = 
\begin{bmatrix}
7043.9575 \\
4357900.1595
\end{bmatrix} 
$$

### Corrección $ \delta $
$$
\delta = 
\begin{bmatrix}
-0.0005456 \\
-2.231959 \times 10^{-7}
\end{bmatrix} 
$$

### Parámetros ajustados
$$
\hat{a} = -0.0005456
$$
$$
\hat{b} = 0.9999998
$$

### Factor de varianza a posteriori de peso unitario $ S_0^2 $
$$
S_0^2 = 1.9428
$$

### Matriz cofactor $ Q_x $
$$
Q_x = 
\begin{bmatrix}
3.1280 \times 10^{-7} & -3.8040 \times 10^{-10} \\
-3.8040 \times 10^{-10} & 6.6609 \times 10^{-13}
\end{bmatrix} 
$$

### Matriz de covarianza de los parámetros $ C_x $
$$
C_x = 
\begin{bmatrix}
6.0771 \times 10^{-7} & -7.3905 \times 10^{-10} \\
-7.3905 \times 10^{-10} & 1.2941 \times 10^{-12}
\end{bmatrix} 
$$

Estos resultados muestran los parámetros ajustados $ a $ y $ b $ y sus respectivas matrices de covarianza y cofactor, siguiendo el procedimiento de mínimos cuadrados ajustados.